tests/test_cli.py

```diff
@@ -428,31 +430,37 @@ def test_all_species_genetic_maps_help(self):
 
 class TestWriteCitations(unittest.TestCase):
     """
-    Make sure all models can write citation information.
+    Make sure all engines, models, etc. can write citation information.
     """
-    def test_model(self):
-        contig = stdpopsim.Contig()
-        species = stdpopsim.get_species("homsap")
-        model = species.get_model("ooa_3")
-        engine = stdpopsim.get_default_engine()
-        stdout, stderr = capture_output(
-                cli.write_citations, engine, model, contig)
-        self.assertEqual(len(stdout), 0)
-        # TODO Parse out the output for the model and check that the text is
-        # in there.
-        self.assertGreater(len(stderr), 0)
-
-    def test_genetic_map(self):
-        species = stdpopsim.get_species("homsap")
-        contig = species.get_contig("chr22", genetic_map="HapmapII_GRCh37")
-        model = stdpopsim.PiecewiseConstantSize(species.population_size)
-        engine = stdpopsim.get_default_engine()
-        stdout, stderr = capture_output(
-                cli.write_citations, engine, model, contig)
-        self.assertEqual(len(stdout), 0)
-        # TODO Parse out the output for the model and check that the text is
-        # in there.
-        self.assertGreater(len(stderr), 0)
+    def test_citations(self):
+        for engine, species in itertools.product(
```

Comment:
Nice. I'm not sure we need to take the products here though. Perhaps we could split this into set of discrete tests, ``test_engine`` and ``test_species``, ``test_genetic_maps`` etc. There's two problems with taking products I think (1) it makes things slow, at some point, as your testing a lot of stuff that doesn't change, and (2) it can be hard to see what's after breaking when a test fails inside in nested loops. Having discrete tests (that uses common verification code) is a nice pattern for this.

Reply:
I agree that there will be too many tests. I'm not sure what choices to make here regarding which engines, species, genetic_maps, and models, should be checked.

As for your point (2), I don't think this too bad, as the assertion message indicates the specific thing that fails (I changed the code slightly to make this more obvious). Eg. the test currently fails like this:

```
======================================================================
FAIL: test_citations (tests.test_cli.TestWriteCitations)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grg/src/stdpopsim/tests/test_cli.py", line 449, in test_citations
    self.check_citations(engine, species, genetic_map, model, stderr)
  File "/home/grg/src/stdpopsim/tests/test_cli.py", line 461, in check_citations
    self.assertTrue(citation.author in stderr, msg=assert_msg)
AssertionError: False is not true : genetic_map: HapmapII_GRCh37
```

Reply:
For unit tests I think it's reasonable to pick one thing to vary, and keep everything else constant unless there are definitely tricky interactions between the things we're testing. In this case, there aren't any interactions between, say, genetic maps and simulation models, so there's no point in doing a cross product there.

Reply:
Ok, done. I also made the assertion failure messages more informative. Current failures are:
```
AssertionError: False is not true : genetic map citation not written for HapmapII_GRCh37
```
and
```
AssertionError: False is not true : generation time citation not written for TestSpecies
```